rfcs/text/0019_lifecycle_preboot.md

```diff
@@ -0,0 +1,246 @@
+
+Serving a proper Kibana application on the root route of the preboot HTTP server implies that we'll also have a chance to replace the static `Kibana server is not ready yet` string with a more helpful and user-friendly application.
+
+### 3.2.9 Preboot service
+
+To support interactive applications at the `preboot` stage we should allow preboot plugins to pause Kibana startup sequence. This functionality will be exposed by the new Preboot service, and will be available to the preboot plugins via `CorePrebootSetup`. Preboot plugins will be able to provide a promise to hold a boot sequence for as much as needed, and also let Kibana know if it has to reload configuration before it enters the `setup` stage.
+
+```ts
+export interface PrebootServiceSetup {
+  readonly isBootOnHold: () => boolean;
+  readonly holdBootUntilResolved: (
+    promise: Promise<{ shouldReloadConfig: boolean } | undefined>
+  ) => void;
+}
+```
+
+Internal Preboot service contract will also expose `waitUntilCanBoot` method that bootstrap process can use to know when it can proceed to `setup` and `start` stages:
+
+```ts
+// NOTE(azasypkin): some existing properties have been omitted for brevity.
+export interface InternalPrebootServiceSetup {
+  readonly waitUntilCanBoot: () => Promise<{ shouldReloadConfig: boolean }>;
+}
+```
+
+### 3.2.10 Bootstrap
+
+We'll update Kibana bootstrap sequence to include `preboot` stage and to conditionally reload configuration before proceeding to `setup` and `start` stages:
+
+```ts
+// NOTE(azasypkin): some functionality and checks have been omitted for brevity.
+const { preboot } = await root.preboot();
+
+const { shouldReloadConfig } = await preboot.waitUntilCanBoot();
+if (shouldReloadConfig) {
+  await reloadConfiguration('pre-boot request');
+}
+
+await root.setup();
+await root.start();
+```
+
+It's not yet clear if we need to adjust the base path proxy to account for this new lifecycle stage (see [unresolved question 3](#83-development-mode-and-basepath-proxy)).
+
+# 4. Drawbacks
+
+The main drawback is that proposed changes affect quite a few Kibana Core services that may impose a risk of breaking something in the critical parts of Kibana.
```

Comment:
What if we have a requirement to support configuration changes from `Stack Management` UI? We will have to remove `preboot` lifecycle support?

Reply:
Can you please elaborate on this a bit? I'm not sure I'm following.

Reply:
I meant that if we had a requirement to support configuration changes from Stack Management UI, Kibana wouldn't leverage the `preboot` phase logic. However, as you said in another comment `if a user has already configured ES connection, we assume they know what they are doing no matter if ES is available or not.` This sounds like a bit different use case, so you can ignore the question.

Reply:
Ah, I see now, thanks. Yeah, to be able to handle configuration changes from the Stack Management UI "at run-time" we'll likely need to widely adopt that reactive approach Josh mentioned in https://github.com/elastic/kibana/pull/99318#pullrequestreview-682859226. Hopefully we can get there at some point.